### PR TITLE
Add deleteAsync and clearBlocking method for improving performance

### PR DESCRIPTION
**Pull request by itosho**
About this issue: https://github.com/cakephp/cakephp/issues/16360

### CODE REVIEW

File: src/Cache/Cache.php [hunk 1]
```diff
@@ -463,6 +489,18 @@ public static function clear(string $config = 'default'): bool
         return static::pool($config)->clear();
     }
 
+    /**
+     * Delete all keys from the cache asynchronously.
+     * But, it is not a fully asynchronous method because keys selection portion is synchronous.
```
Comment (markstory) [inline src/Cache/Cache.php]: Should this comment mention that this approach is 'blocking' not asynchronous?

File: src/Cache/CacheEngine.php [hunk 2]
```diff
@@ -229,6 +229,28 @@ public function has($key): bool
...
 
+    /**
+     * Delete a key from the cache asynchronously
+     *
+     * @param string $key Identifier for the data
+     * @return bool True if the value was successfully unlinked, false if it didn't exist or couldn't be unlinked
+     */
+    public function deleteAsync($key): bool
+    {
+        return $this->delete($key);
```
Comment (markstory) [inline src/Cache/CacheEngine.php]: Should this method raise an exception? It doesn't feel great that the method name has 'async' and the default behavior is to do a blocking delete.

File: src/Cache/Cache.php (hunk 1 above)
Comment (itosho) [reply src/Cache/Cache.php]: Sorry, this PR was still a DRAFT and I will mention it to you when it is READY.

File: src/Cache/CacheEngine.php (hunk 2 above)
Comment (itosho) [reply src/Cache/CacheEngine.php]: Sorry, this PR was still a DRAFT and I will mention it to you when it is READY.

**Comment by itosho** [comment]
@markstory Please review 🙏 
(The codecov/project check is failing, but it seems difficult to fix.)

File: src/Cache/Cache.php [hunk 3]
```diff
@@ -463,6 +489,17 @@ public static function clear(string $config = 'default'): bool
         return static::pool($config)->clear();
     }
 
+    /**
+     * Blocking delete all keys from the cache
```
Comment (markstory) [inline src/Cache/Cache.php]:
```suggestion
     * Blocking delete of all keys from the cache
```

File: src/Cache/CacheEngine.php [hunk 4]
```diff
@@ -279,13 +280,40 @@ abstract public function decrement(string $key, int $offset = 1);
...
+     *
+     * Empty method and always raises an exception.
+     * It needs to be implemented in a concrete cache engine.
+     *
+     * @param string $key Identifier for the data
+     * @return bool True if the value was successfully deleted, false if it didn't exist or couldn't be removed
+     */
+    public function deleteAsync($key): bool
+    {
+        throw new NotImplementedException('There is no implementation of a concrete cache engine.');
```
Comment (markstory) [inline src/Cache/CacheEngine.php]:
```suggestion
        throw new NotImplementedException('The deleteAsync() method is not supported by' . static::class . '. Used delete() instead.');
```

We can like to give more direction in our error messages where possible.

File: src/Cache/CacheEngine.php [hunk 5]
```diff
@@ -279,13 +280,40 @@ abstract public function decrement(string $key, int $offset = 1);
...
+     * Blocking delete all keys from the cache
+     *
+     * Empty method and always raises an exception.
+     * It needs to be implemented in a concrete cache engine.
+     *
+     * @return bool True if the cache was successfully cleared, false otherwise
+     */
+    public function clearBlocking(): bool
+    {
+        throw new NotImplementedException('There is no implementation of a concrete cache engine.');
```
Comment (markstory) [inline src/Cache/CacheEngine.php]:
```suggestion
        throw new NotImplementedException('Fast blocking clears are not supported by ' . static::class . '. Use clear() instead.');
```

**Review by markstory — COMMENTED** [review]
Looking good so far.

File: src/Cache/CacheEngine.php [hunk 6]
```diff
@@ -17,6 +17,7 @@
 namespace Cake\Cache;
 
 use Cake\Core\InstanceConfigTrait;
+use Cake\Http\Exception\NotImplementedException;
```
Comment (ADmad) [inline src/Cache/CacheEngine.php]: Using a HTTP exception doesn't make sense here. You can add a new exception class under the Cache/Exception namespace which extends CakeException.

Comment (itosho) [reply src/Cache/CacheEngine.php]: OK, I'll try it 💪

Comment (itosho) [reply src/Cache/CacheEngine.php]: fixed https://github.com/cakephp/cakephp/pull/16478/commits/676c65d3048c1f6a8eaa5f9c17c65939b103f6bf

File: src/Cache/CacheEngine.php [hunk 7]
```diff
@@ -291,7 +291,8 @@ abstract public function delete($key): bool;
      */
     public function deleteAsync($key): bool
     {
-        throw new NotImplementedException('The deleteAsync() method is not supported by' . static::class . '. Used delete() instead.');
+        $message = 'The deleteAsync() method is not supported by' . static::class . '. Used delete() instead.';
```
Comment (ADmad) [inline src/Cache/CacheEngine.php]:
```suggestion
        $message = 'The deleteAsync() method is not supported by ' . static::class . '. Use delete() instead.';
```

File: src/Cache/CacheEngineInterface.php [hunk 8]
```diff
@@ -66,4 +66,19 @@ public function decrement(string $key, int $offset = 1);
...
      */
     public function clearGroup(string $group): bool;
+
+    /**
+     * Delete a key from the cache asynchronously
+     *
+     * @param string $key Identifier for the data
+     * @return bool True if the value was successfully deleted, false if it didn't exist or couldn't be removed
+     */
+    public function deleteAsync(string $key): bool;
```
Comment (ADmad) [inline src/Cache/CacheEngineInterface.php]: We can't change the interface in a minor release. Instead add `@method` annotations for these new methods for the time being.

File: src/Cache/CacheEngine.php [hunk 9]
```diff
@@ -279,13 +280,42 @@ abstract public function decrement(string $key, int $offset = 1);
...
+    /**
+     * Delete a key from the cache asynchronously
+     *
+     * Empty method and always raises an exception.
+     * It needs to be implemented in a concrete cache engine.
+     *
+     * @param string $key Identifier for the data
+     * @return bool True if the value was successfully deleted, false if it didn't exist or couldn't be removed
+     */
+    public function deleteAsync($key): bool
```
Comment (ADmad) [inline src/Cache/CacheEngine.php]:
```suggestion
    public function deleteAsync(string $key): bool
```

File: src/Cache/Engine/RedisEngine.php [hunk 10]
```diff
@@ -227,6 +227,21 @@ public function delete($key): bool
...
 
+    /**
+     * Delete a key from the cache asynchronously
+     *
+     * Just unlink a key from the cache. The actual removal will happen later asynchronously.
+     *
+     * @param string $key Identifier for the data
+     * @return bool True if the value was successfully deleted, false if it didn't exist or couldn't be removed
+     */
+    public function deleteAsync($key): bool
```
Comment (ADmad) [inline src/Cache/Engine/RedisEngine.php]:
```suggestion
    public function deleteAsync(string $key): bool
```

**Comment by ADmad** [comment]
Are there any other cache engines which implement these new methods?

**Comment by itosho** [comment]
> Are there any other cache engines which implement these new methods?

I'm not completely sure, but I don't think it is implemented in memcached.
ref: https://github.com/memcached/memcached/wiki/Commands

**Comment by ADmad** [comment]
Adding the new methods to `Cache` and `CacheEngineInterface` doesn't feel right to me if there's going to be only 1 engine implementating it.

We could have them for the RedisEngine only and they could be used as `Cache::pool('foo')->method()`.

**Comment by itosho** [comment]
> Adding the new methods to Cache and CacheEngineInterface doesn't feel right to me if there's going to be only 1 engine implementating it.

Yes.
I added these methods to the interface because I was not aware of the conventions for cache implementation.

I will look into other cache engines and if Redis is the only one, I will fix them as you advised.

**Comment by itosho** [comment]
Based on my research, the implementation of asynchronous deletion for each cache engine is as follows.

- Array: N/A
  - only [unset](https://www.php.net/manual/en/function.unset.php)
- Apcu: N/A
  - only [apcu-delete](https://www.php.net/manual/en/function.apcu-delete.php)
- File: N/A
  - only [unlink](https://www.php.net/manual/en/function.unlink.php)(not asynchronous) 
- Memcached: N/A
  - only [delete](https://github.com/memcached/memcached/wiki/Commands#delete ) 
- Null: N/A
- Redis: Available
  - [del](https://redis.io/commands/del/) and [unlink](https://redis.io/commands/unlink/)
- Win: N/A
  - only [wincache-ucache-delete](https://www.php.net/manual/en/function.wincache-ucache-delete.php)

**Comment by itosho** [comment]
I will create another PR since the implementation has changed from the beginning.

**Pull request closed without merging**